### PR TITLE
Allow creation of sign-less init packets

### PR DESCRIPTION
**Pull request by IvanSanchez**
TL;DR: users can create zip files without signing them

long version:
- Unit tests for the low-ish level protobuf funcionality (ensure that previous functionality is maintained)
- Allow for low-ish level instantiation of init packets protobufs with no signature
- Stop high-ish functionality (argument parser) from requiring a `--key` option
- `nrfutil pkg display` can deal with both signed and unsigned init packet protobufs
- Hook all the previous bits together
- Some minor changes and typo fixes on help strings

### CODE REVIEW

**Comment by IvanSanchez** [comment]
Note how the protobuf for the unsigned init packet is smaller, and a signed one contains the unsigned one, as expected from the protobuf definition.

image

File: tests/protobuf/__main__.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+                 hash_type = HashTypes.SHA256,
+                 dfu_type = DFUType.APPLICATION,
+                 is_debug=False,
+                 fw_version=0xffffffff,
+                 hw_version=0xffffffff,
+                 sd_size=0,
+                 app_size=1234,
+                 bl_size=0,
+                 sd_req=[0xffffffff]
+                )
```
Comment (chunfantasy) [inline tests/protobuf/__main__.py]: Incorrect indent.

File: tests/protobuf/__main__.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
```
Comment (chunfantasy) [inline tests/protobuf/__main__.py]: Same as above.

File: tests/protobuf/__main__.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
```
Comment (chunfantasy) [inline tests/protobuf/__main__.py]: Same as above.

File: tests/protobuf/__main__.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+        i.set_signature(b"signature bytes go here", SigningTypes.ECDSA_P256_SHA256)
+
+        protobuf_bytes = i.get_init_packet_pb_bytes(); # Bytes for the whole Packet
+        hex_bytes = protobuf_bytes.encode('hex_codec');
+
+        self.assertEqual(hex_bytes, "12450a280801122408ffffffff0f10ffffffff0f1a" "05ffffffff0f20002800300038d209420408031200480010001a177369676e61747572"
```
Comment (chunfantasy) [inline tests/protobuf/__main__.py]: Incorrect indent.

File: tests/protobuf/__main__.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+sys.path.append(
+    os.path.normpath(
+        os.path.join(
+            os.path.dirname(__file__), '..', '..' # , 'nordicsemi'
+        )
+    )
+)
+
+#from nordicsemi.version import NRFUTIL_VERSION
```
Comment (chunfantasy) [inline tests/protobuf/__main__.py]: Commented out code should be removed.

File: tests/protobuf/__main__.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    #def test_isupper(self):
+        #self.assertTrue('FOO'.isupper())
+        #self.assertFalse('Foo'.isupper())
+
+    #def test_split(self):
+        #s = 'hello world'
+        #self.assertEqual(s.split(), ['hello', 'world'])
+        ## check that s.split fails when the separator is not a string
+        #with self.assertRaises(TypeError):
+            #s.split(2)
```
Comment (chunfantasy) [inline tests/protobuf/__main__.py]: Commented out code should be removed.

File: tests/protobuf/__main__.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                 app_size=1234,
+                 bl_size=0,
+                 sd_req=[0xffffffff]
+                )
+
+        protobuf_bytes = i.get_init_packet_pb_bytes(); # Bytes for the whole Packet
+        hex_bytes = protobuf_bytes.encode('hex_codec');
+
+        self.assertEqual(hex_bytes, "0a280801122408ffffffff0f10ffffffff0f1a"
+        "05ffffffff0f20002800300038d2094204080312004800")
```
Comment (chunfantasy) [inline tests/protobuf/__main__.py]: Or maybe here also?

File: nordicsemi/__main__.py [hunk 8]
```diff
@@ -350,15 +350,15 @@ def display(key_file, key, format, out_file):
             kfile.write(kstr)
 
 
-@cli.group(short_help='Generate a Device Firmware Update package.')
+@cli.group(short_help='Display or generate a DFU package (zip file).')
```
Comment (chunfantasy) [inline nordicsemi/__main__.py]: I dont see any updates for displaying a package.
Is it a good description?

Comment (IvanSanchez) [reply nordicsemi/__main__.py]: Yes, it's good. This is the docstring for the `nrfutil pkg` command.

The `nrfutil pkg` command has two subcommands: `nrfutil pkg display` and `nrfutil pkg generate`, so this is actually a correction.

Comment (chunfantasy) [reply nordicsemi/__main__.py]: 👍

File: tests/protobuf/__main__.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+from nordicsemi.dfu.init_packet_pb import InitPacketPB, HashTypes, DFUType, SigningTypes
+
+class TestStringMethods(unittest.TestCase):
+
+    def test_construct_from_empty_string(self):
+        """
+        Raises an error when the trying to construct a init packet from a
+        empty buffer
```
Comment (mrodem) [inline tests/protobuf/__main__.py]: Typos:
* s/when the trying/when trying/
* s/from a empty/from an empty/

File: tests/protobuf/__main__.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        self.assertRaisesRegexp(
+            RuntimeError,
+            "app_size is not set. It must be set when type is APPLICATION",
+            InitPacketPB,
+            from_bytes = "");
+
+
+    def test_construct_from_params(self):
+        """
+        Gracefully constructs a init packet protobuffer from parameters without bytes
```
Comment (mrodem) [inline tests/protobuf/__main__.py]: s/constructs a/constructs an/

File: tests/protobuf/__main__.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+        self.assertEqual(hex_bytes, "08ffffffff0f10ffffffff0f1a05ffffffff0f20002800"
+                         "300038d2094204080312004800")
+
+        #print i.get_init_packet_pb_bytes();
+
+
+    def test_construct_from_params_and_sign(self):
+        """
+        Gracefully constructs a init packet protobuffer from parameters without bytes,
```
Comment (mrodem) [inline tests/protobuf/__main__.py]: s/constructs a/constructs an/

File: tests/protobuf/__main__.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        hex_bytes = protobuf_bytes.encode('hex_codec');
+
+        self.assertEqual(hex_bytes, "12450a280801122408ffffffff0f10ffffffff0f1a"
+        "05ffffffff0f20002800300038d209420408031200480010001a177369676e61747572"
+        "6520627974657320676f2068657265")
+
+
+    def test_construct_from_params_and_not_sign(self):
+        """
+        Gracefully constructs a init packet protobuffer from parameters without bytes,
```
Comment (mrodem) [inline tests/protobuf/__main__.py]: s/constructs a/constructs an/

File: tests/protobuf/__main__.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            sd_req=[0xffffffff]
+            )
+
+        protobuf_bytes = i.get_init_command_bytes(); # bytes only for the InitCommand
+        hex_bytes = protobuf_bytes.encode('hex_codec');
+
+        self.assertEqual(hex_bytes, "08ffffffff0f10ffffffff0f1a05ffffffff0f20002800"
+                         "300038d2094204080312004800")
+
+        #print i.get_init_packet_pb_bytes();
```
Comment (mrodem) [inline tests/protobuf/__main__.py]: Remove.

File: tests/protobuf/__main__.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
+
+import unittest
+import os, sys
+
+sys.path.append(
+    os.path.normpath(
+        os.path.join(
+            os.path.dirname(__file__), '..', '..' # , 'nordicsemi'
```
Comment (mrodem) [inline tests/protobuf/__main__.py]: Remove `# , 'nordicsemi'`?

File: nordicsemi/__main__.py [hunk 15]
```diff
@@ -571,10 +573,11 @@ def generate(zipfile,
...
 
-    signer = Signing()
-    default_key = signer.load_key(key_file)
-    if default_key:
-        display_sec_warning()
+    if key_file is not None:
+        signer = Signing()
+        default_key = signer.load_key(key_file)
+        if default_key:
+            display_sec_warning()
```
Comment (mrodem) [inline nordicsemi/__main__.py]: Currently, a warning is displayed when signing with the default key or when using the debug flag (e.g. "do not do this in production"). Perhaps a warning should be displayed when generating unsigned packages as well?

Comment (IvanSanchez) [reply nordicsemi/__main__.py]: Good point. I did not check those warnings.

**Pull request merged**